Flexible localization options.

matthew-macgregor opened the pull request:
## Summary

The existing localization code relies upon Locale.getDefault(), which uses the Locale of the host JVM. While this can be changed programmatically, the change applies globally for the JVM. This is fine for client-side use, but it presents issues when used in a server-side environment where each request may need to use a different locale concurrently. 

After attempting several workarounds and investigating code, I realized that this use case hasn't been addressed and decided to attempt a solution.

You can see a discussion here:

[StackOverflow: How to change the Locale of epubcheck when using it as a library](http://stackoverflow.com/questions/33586984/how-to-change-the-locale-of-epubcheck-when-using-it-as-a-library/33588298)
### Goals
- Allow clients to pass a locale to the EpubCheck class.
- Fall back to the default locale if no explicit locale is provided.
- Preserve caching of the resource-hungry localized strings.
- Use lazy initialization of messages wherever possible. 
- Minimize disturbance of top-level interfaces.
### Usage

Usage of the new functionality is simple:

```
Locale l = new Locale("es");
EpubCheck epubCheck = new EpubCheck(file);
epubCheck.setLocale(l); // new
epubCheck.doValidate();

// Validación usando la versión de reglas epub 2.0.1.
```
### Changes
- Refactored `com.adobe.epubcheck.util.Messages` to cache an instance of itself for each locale, as they are requested. Updated all the tests.
- Refactored `com.adobe.epubcheck.messages` to use a similar pattern. The overridden messages added some complications. As an attempt to modularize, I broke `MessageDictionary.java` into several classes, which allowed me to cache what I wanted to cache (default messages and severities) while keeping the override functionality intact. 
- I added `setLocale()` to the `Report` interface. This will break any clients that are passing custom reports. The code works without this with minor modifications (since we have access via the `MasterReport`) but it seemed more consistent to keep this on the interface. It could easily be removed to avoid the breaking change. 
- Updated tests and added new tests for the additional functionality. 
### Notes & Limitations
- There are currently two Message API's in the code base: one which handles epubcheck-related messages and the other which handles "utility" messages. I left this division intact, but propose that the util messages should be merged into the more robust MessageDictionary structure in some way.
- Since localization is tied to the MasterReport, clients that provide a custom Report object will find it difficult to benefit from localization without inheriting from MasterReport. This is a problem with both the existing code and my changes. I didn't address it.

Thanks very much, I'm looking forward to feedback.

rdeltour commented:
Thanks for the PR, which at first sight looks useful and well documented :+1: 
The diff is rather significant and I can't say when I'll have the time to properly review, but I'm assigning it to myself for a next release anyway.

matthew-macgregor commented:
Thanks for tackling the pull request! I'm happy to answer questions or do a walkthrough when the time comes for a review.

Regarding the number of changes, it turned into a bigger changeset than I initially expected. Here are a few notes that might help code review:

The overridden files functionality caused some challenges with per-locale static caching of messages. This led me to split up the `MessageDictionary` class, which was handling default messages, loading of override files, overridden messages, dumping message files and severities. These are now separate entities. The top level object, `MessageDictionary`, remains but is now an interface.  Most of the code in the implementations is unchanged but reorganized into new, smaller files. I didn't like the alternative, which was to make `MessageDictionary` responsible for still more functionality. In any case, you can refer back to that original `MessageDictionary` class, most of the code is untouched, just reorganized.

I was worried about side effects with the override files. The code to load the files is unchanged but the code to use the data is new. I let the existing command line tests (which broke initially) help guide me to the correct behavior. I added a new test for just the new `OverriddenMessageDictionary` class with my assumptions. You may wish to review that test to make sure that the override rules I followed make sense. 

In the original, overrides literally overwrote the original (static) message in memory. This obviously caused problems for the localization feature. Now the override is handled as a fallback: take the overridden message if available, or fallback to the default localization if not. It amounts to the same thing, but the logic was implicit before and explicit now. I separated all of this into its own implementation, so it's clear it's doing something different.

tofi86 commented:
I would like to give this a go and merge the PullRequest soon as it would fix an issue with my EpubCheck GUI: https://github.com/paginagmbh/EPUB-Checker/issues/23

@matthew-macgregor there's a lot of refactoring going on and I didn't check all of it yet, but it looks very promising and well tested. I tried to use it in my GUI app and it seems to work pretty well. There was only one small change I had to make in my custom report (https://github.com/paginagmbh/EPUB-Checker/blob/master/src/main/java/de/paginagmbh/epubchecker/paginaReport.java#L67) . But as it inherits from `DefaultReportImpl` which inherits from `MasterReport` I was lucky for not having too many changes.

@rdeltour however, as @matthew-macgregor stated, the changes to the `Report` class will probably break existing custom Reports which doesn't inherit from `MasterReport`. Would you even consider this PR then for a bugfix release 4.0.3? Or shall we better call the next version 4.1.0 then to make clear that it's not only a bugfix release and may brake something? Or would you like to wait even longer? 5.0?

@matthew-macgregor Is there any chance you could rebase your PR on current `master` with attention to the changes made in 70e990d for #711 and the now-merged PR #712?
This would speed up testing and merging of your awesome contribution!

tofi86 commented:
Probably we can also add a language switch to the commandline interface to either suppress translation (default to `en` messages) or request a specific (supported) localization.

refs #600

matthew-macgregor commented:
@tofi86 Sorry about the slow response. Thanks for considering this contribution. 

I'll be glad to rebase [#650](https://github.com/IDPF/epubcheck/pull/650) on current master with attention to the other PR's that you mentioned ([#711](https://github.com/IDPF/epubcheck/issues/711) & [#712](https://github.com/IDPF/epubcheck/pull/712)). 

Would you like me to also look at the command line language switch?

tofi86 commented:
> @tofi86 Sorry about the slow response. Thanks for considering this contribution.

Sorry for OUR slow response (it's been a year now... 🙁) 

> I'll be glad to rebase #650 on current master with attention to the other PR's that you mentioned (#711 & #712).
> 
> Would you like me to also look at the command line language switch?

Thanks a lot!

Well, if you +1 my idea of adding a language option to the cmdl interface, that would be awesome if you'd take a look into that as well!

tofi86 commented:
@matthew-macgregor any idea when you want to start working on this? I absolutely don't want to rush you with this—just asking for a rough schedule as it would possibly be the best to merge #723 and #731 first as they also have changes to the message files and I'm still waiting for a _LGTM_ from @rdeltour for those...

tofi86 commented:
> Would you like me to also look at the command line language switch?

@matthew-macgregor A `--lang <lang>` argument has also be requested in #498.
I suggest we better add this instead of a "don't translate" switch (#600)...

matthew-macgregor commented:
@tofi86 Hi, I see you asked for an estimate weeks ago. I'm sorry, I've been heads down in a work project and barely checking my personal accounts. I'm going to look at this today and can provide a better idea.

matthew-macgregor commented:
@tofi86 I rebased (dry run) against current master. Everything looks straightforward, I've re-familiarized myself with the localization changes and all tests are passing. I've also taken a look at [#723](https://github.com/IDPF/epubcheck/pull/723) and [#731](https://github.com/IDPF/epubcheck/pull/731) and those should also be easy to integrate. Just give me the nod when those are merged and I'll do the process again. It should be quick.

I've got a couple of notes:

1. If an explicit Locale isn't provided I'm using the default Locale, which is consistent with how the library worked prior to my changes. Due to backwards-compatibility, this is probably the correct implementation for this release. However, it might provide better sanity to have EpubCheck set its own default locale rather than relying on the JVM default (which seems to cause confusion). 
2. Is it easier for you if I provide the command line language option as a separate PR or would you rather that I integrate it with [#650](https://github.com/IDPF/epubcheck/pull/650)?

tofi86 commented:
> @tofi86 I rebased (dry run) against current master. Everything looks straightforward, I've re-familiarized myself with the localization changes and all tests are passing. I've also taken a look at #723 and #731 and those should also be easy to integrate. Just give me the nod when those are merged and I'll do the process again. It should be quick.

Great, thanks for giving this a shot! And also thanks for having a look at the outstanding PullRequests. I could merge #723 soon, but would like to have a review first for #731.


> If an explicit Locale isn't provided I'm using the default Locale, which is consistent with how the library worked prior to my changes. Due to backwards-compatibility, this is probably the correct implementation for this release. However, it might provide better sanity to have EpubCheck set its own default locale rather than relying on the JVM default (which seems to cause confusion).

What do you suggest? Using `en_US` as default locale?


> Is it easier for you if I provide the command line language option as a separate PR or would you rather that I integrate it with #650?

If you can integrate with this PR it would be great!

tofi86 commented:
@matthew-macgregor PullRequest #723 and #731 with pending changes to the message files have successfully been merged. You can go ahead and rebase on current master.

Thanks for your patience and your contribution! 🎉

tofi86 commented:
@matthew-macgregor Any chance to give this a second rebase? Next EpubCheck release (v4.1) is currently scheduled for end of august and I'd really like to see this PR included!

matthew-macgregor commented:
@tofi86 Excellent, I'll do it this weekend.

matthew-macgregor commented:
I have updated the pull request with the changes, including an option for the CLI. Here's the summary, and things to be aware of:

- I used `--locale` instead of `--lang`. After some reflection, this seemed more intuitive and to better reflect what's actually happening under the covers.
- In the case of no Locale, the library will fall back to the default JVM Locale, which is consistent with previous behavior. In the case of an invalid Locale, or a Locale which has no translation, messages will output in English.
- After some waffling, I decided not to attempt to validate the Locale. This seems to follow the pattern of Locale.forLanguageTag(). The practical result is that an invalid Locale will cause messages to output in English as noted above. Alternatively, we could use the commons-lang to validate the language tag, but it didn't seem worth the added dependency, nor did it seem worth rolling our own validation.
- I added localized error and help messages for the CLI using **Google Translate**, so they're probably pretty terrible. If we've got native speakers who can do these properly, that would be great. 
- I didn't attempt the Japanese translation at all, I'm afraid, but left an English placeholder in that file. Again, hopefully we have a native speaker who can provide that translation.

Thanks!

murata0204 commented:
I believe that Keio APL can take care of the Japanese translation.  Should we start after this pull request is incorporated?

tofi86 commented:
@murata0204 that's great news, thanks for taking care of that!
However, as stated in the requested changes, we don't translate the non-english message property files in the repository but in the Transifex translation project for epubcheck, following the rules written on this wiki page: https://github.com/IDPF/epubcheck/wiki/Translating
So when this PR is finally merged, you or your colleagues could create an account with Transifex and start translating the missing Japanese messages. Not only for this PR but for all the changes within this Milestone.

Before the 4.1.0 release, I will call for general translation updates anyways.

matthew-macgregor commented:
@tofi86 I've reverted the changes to the `messages_*.properties` files and updated the PR. Thanks!

tofi86 commented on src/main/resources/com/adobe/epubcheck/util/messages.properties:
```diff
@@ -20,7 +20,7 @@ argument_needed=At least one argument expected
 mode_argument_expected=Type omitted from the -mode argument.
 profile_argument_expected=Profile name omitted after the -profile argument.
-profile_unknown=Unknown profile '%1$s', using default validation profile.
+profile_unknown=Unknown profile '%1$s', usingÂ default validation profile.
```
there seems to be a unicode character glitch - can you have a look @matthew-macgregor ?

tofi86 replied:
@rdeltour fixed this with commit 91ccd2f62fa944b44644be62aac098b06accccad

tofi86 commented:
When building your branch with maven and Java 7 on my Windows system with locale `de_DE` I get the following test errors:

```
Failed tests:
  ensureOverridenMessages(com.adobe.epubcheck.test.OverriddenMessageDictionaryTest):
    Message should have been overridden.
  ensureNullIsDefault(com.adobe.epubcheck.test.UtilMessage_LocalizationTest):
    Null locale should use (host) default locale. expected:<en_DE> but was:<en_US>
  ensureDefault(com.adobe.epubcheck.test.UtilMessage_LocalizationTest):
    Default constructor should use the (host) default locale. expected:<en_DE> but was:<en_US>
```
<details>
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.adobe.epubcheck.api.Epub20CheckExpandedTest
Tests run: 22, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5.947 sec
Running com.adobe.epubcheck.api.Epub20CheckTest
Spine element not found within package root document ops/manifest.opf
Spine element not found within package root document EPUB/manifest.opf
Spine element not found within package root document OEBPS/content.opf
Tests run: 31, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.086 sec
Running com.adobe.epubcheck.api.Epub30CheckExpandedTest
Tests run: 147, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 8.758 sec
Running com.adobe.epubcheck.api.Epub30CheckTest
Tests run: 44, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.225 sec
Running com.adobe.epubcheck.cli.CLITest
No errors or warnings detected.
Messages: 0 fatal / 0 errors / 0 warnings / 0 info

epubcheck completed
Tests run: 17, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.291 sec
Running com.adobe.epubcheck.nav.NavCheckerTest
Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.045 sec
Running com.adobe.epubcheck.ocf.OCFCheckerTest
Tests run: 25, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.411 sec
Running com.adobe.epubcheck.ocf.OCFFilenameCheckerTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 sec
Running com.adobe.epubcheck.opf.MetadataSetTest
Tests run: 13, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 sec
Running com.adobe.epubcheck.opf.OPFCheckerTest
Tests run: 126, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.636 sec
Running com.adobe.epubcheck.ops.OPSCheckerTest
Tests run: 100, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.384 sec
Running com.adobe.epubcheck.overlay.OverlayCheckerTest
Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.024 sec
Running com.adobe.epubcheck.test.api_Test
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.076 sec
Running com.adobe.epubcheck.test.command_line_Test
Tests run: 19, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.879 sec
Running com.adobe.epubcheck.test.css_Test
Tests run: 17, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.55 sec
Running com.adobe.epubcheck.test.dtBook_Test
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.289 sec
Running com.adobe.epubcheck.test.encryption_Test
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.379 sec
Running com.adobe.epubcheck.test.MessageDictionary_LocalizationTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running com.adobe.epubcheck.test.opf_Test
Tests run: 30, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.526 sec
Running com.adobe.epubcheck.test.OverriddenMessageDictionaryTest
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec <<< FAILURE!
ensureOverridenMessages(com.adobe.epubcheck.test.OverriddenMessageDictionaryTest)  Time elapsed: 0.002 sec  <<< FAILURE!
java.lang.AssertionError: Message should have been overridden.
        at org.junit.Assert.fail(Assert.java:93)
        at org.junit.Assert.assertTrue(Assert.java:43)
        at com.adobe.epubcheck.test.OverriddenMessageDictionaryTest.ensureOverridenMessages(OverriddenMessageDictionaryTest.java:29)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:45)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:15)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:42)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:263)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:68)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:47)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:231)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:60)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:229)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:50)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:222)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:300)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
        at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
        at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)

Running com.adobe.epubcheck.test.package_Test
Tests run: 21, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.463 sec
Running com.adobe.epubcheck.test.script_Test
Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.384 sec
Running com.adobe.epubcheck.test.single_file_Test
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.017 sec
Running com.adobe.epubcheck.test.toc_Test
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.153 sec
Running com.adobe.epubcheck.test.UtilMessage_LocalizationTest
Tests run: 5, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0 sec <<< FAILURE!
ensureNullIsDefault(com.adobe.epubcheck.test.UtilMessage_LocalizationTest)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: Null locale should use (host) default locale. expected:<en_DE> but was:<en_US>
        at org.junit.Assert.fail(Assert.java:93)
        at org.junit.Assert.failNotEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:128)
        at com.adobe.epubcheck.test.UtilMessage_LocalizationTest.ensureNullIsDefault(UtilMessage_LocalizationTest.java:32)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:45)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:15)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:42)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:20)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:28)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:263)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:68)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:47)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:231)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:60)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:229)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:50)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:222)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:300)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
        at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
        at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)

ensureDefault(com.adobe.epubcheck.test.UtilMessage_LocalizationTest)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: Default constructor should use the (host) default locale. expected:<en_DE> but was:<en_US>
        at org.junit.Assert.fail(Assert.java:93)
        at org.junit.Assert.failNotEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:128)
        at com.adobe.epubcheck.test.UtilMessage_LocalizationTest.ensureDefault(UtilMessage_LocalizationTest.java:23)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:45)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:15)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:42)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:20)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:28)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:263)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:68)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:47)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:231)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:60)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:229)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:50)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:222)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:300)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
        at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
        at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)

Running com.adobe.epubcheck.test.xhtml_Test
Tests run: 17, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.525 sec
Running com.adobe.epubcheck.util.OPFPeekerTest
Tests run: 19, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.008 sec
Running com.adobe.epubcheck.util.PathUtilTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 sec
Running com.adobe.epubcheck.vocab.PrefixParsingTest
Tests run: 13, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 sec
Running com.adobe.epubcheck.vocab.PropertyTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running com.adobe.epubcheck.vocab.VocabTest
Tests run: 14, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 sec
Running org.idpf.epubcheck.util.css.CssInputStreamTest
Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 sec
Running org.idpf.epubcheck.util.css.CssParserTest
Tests run: 120, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 sec
Running org.idpf.epubcheck.util.css.CssScannerTest
Tests run: 127, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 sec

Results :

Failed tests:   ensureOverridenMessages(com.adobe.epubcheck.test.OverriddenMessageDictionaryTest): Message should have been overridden.
  ensureNullIsDefault(com.adobe.epubcheck.test.UtilMessage_LocalizationTest): Null locale should use (host) default locale. expected:<en_DE> but was:<en_US>
  ensureDefault(com.adobe.epubcheck.test.UtilMessage_LocalizationTest): Default constructor should use the (host) default locale. expected:<en_DE> but was:<en_US>

Tests run: 998, Failures: 3, Errors: 0, Skipped: 1

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:07 min
[INFO] Finished at: 2017-08-15T19:15:29+02:00
[INFO] Final Memory: 31M/195M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.12.4:test (default-test) on project epubcheck: There are test failures.
</details>

Can you please have a look at these and try to reproduce?

Please tell me if I can help you debug this issue!

tofi86 commented:
@matthew-macgregor Hi Matthew, unfortunately I have encountered another problem:

```
ERROR (CSS-008) at "waldenfotos.epub/OEBPS/Walden_c1.xhtml" (line 10, col 3):
   An error occurred while parsing the CSS: Die Zeichenkette ".quote" ist an dieser Stelle nicht erlaubt. Erwartet wird eine CSS-Eigenschaft.
```

This message appeared in my EPUB-Checker App after switching the language from `de_DE` to `en_US`.

The first part of the message is in english and is issued by the [EpubCheck message libs](https://github.com/matthew-macgregor/epubcheck/blob/localization-dev/src/main/resources/com/adobe/epubcheck/messages/MessageBundle.properties#L44) while the second part, which comes from the [Jing validation](https://github.com/matthew-macgregor/epubcheck/blob/localization-dev/src/main/resources/org/idpf/epubcheck/util/css/messages_de.properties#L10), is still in german.

I haven't figured out yet what the problem is, but if you could take a closer look at this, that would be nice.

I set the locale in my custom Report clas with `report.setLocale()` and also with `EpubCheck.setLocale()`. This should be enough, shouldn't it?

tofi86 commented:
> The first part of the message is in english and is issued by the EpubCheck message libs while the second part, which comes from the Jing validation, is still in german.

I was able to reproduce this the other way round in Eclipse, outside of my App, on your plain `localization-dev` branch with the following run configuration:

Arguments:
```
--locale de "C:\PATH-TO-EPUB-FILE\waldenfotos.epub"
```

VM arguments:
```
-Duser.language=en -Duser.country=US
```

And my system is `de_DE` by default. So you probably have to switch the locale argument and user lang

Resulting message is german/english mix:

```
ERROR(CSS-008): C:/PATH-TO-EPUB-FILE/waldenfotos.epub/OEBPS/Walden_c1.xhtml(10,3): CSS-Validierungsfehler: Token '.quote' not allowed here, expecting a property name
```

tofi86 commented:
@matthew-macgregor could you probably have another look at the issues I encountered when testing the PR? thanks!

tofi86 commented:
Hi @matthew-macgregor, we're way behind our release schedule. But this is still a top-priority contribution! Could you please check my feedback in the next few weeks so that we can get this working and ship with 4.1.0? That would be awesome! 👍

matthew-macgregor commented:
@tofi86 Thanks for your patience. I took a look this weekend and I believe that I found two issues. I have a solution for the first issue, but I'm not sure the best way to handle the second issue.

First, I discovered that when the default locale is non-English and the `--locale` override is "en" there was a subtle bug. In this case it was looking for a localized properties file `*_en.properties`, and not finding one, it was falling back to the Locale default. I provided empty `*_en.properties` files which resolved that issue.

The second issue is that I missed another set of properties files, and another Messages class in `org.ipdf.epubcheck.util.css`. This is the cause of the weird mixed-language message:  CSS-Validierungsfehler: Token '.quote' not allowed here, expecting a property name. That other Messages class is still using Locale.getDefault(), so when you provided the `-Duser.language=en -Duser.country=US` arguments any messages pulled from those files don't get the new behavior.

I looked through that code with the intent of fixing it. `org.idpf.epubcheck.util.css.Messages` is basically a duplicate of the original `com.adobe.epubcheck.util.Messages`, literally a copy-paste except for the bundle id, so those should ideally be merged anyway. The problem is that static `Messages.get()` calls are littered through the css validation code, including in the overridden Exception classes. Unfortunately, I'll need to inject the locale down through lots of layers (the way it's currently structured, we'll need it in the `CssException` constructor itself). I think this doesn't smell right (but will work), or it needs a more thoughtful refactor. If you've got any ideas on that front, I'd love to hear them.

I'll update the PR soon, I'd like to look at a couple of other things first and add a few more tests. Sorry about the hold up on my end.

tofi86 commented:
Hi Matthew,

thanks for having another look at the issue. Looking forward to your updated solution.

I will also try to take a look at the CSS messages class, but probably @rdeltour can help better with this.

tofi86 commented:
In todays EpubCheck TaskForce Meeting @rdeltour promised to take a look in the next couple of weeks. I feel a bit uncertain or lost on how to decide that topic, @matthew-macgregor :-/

However, we also have a fixed release date now by the end of november and we should try to find a solution for that in the next 20 days...

matthew-macgregor commented:
@tofi86 Status update. I believe I've fixed the issue you noticed for messages originating within the library, although it has forced me to inject the locale through layers that don't really make sense. This feels haphazard and tough to debug, and I just don't like it for hygiene reasons.

I've hit another roadblock. While testing, I still see the mixed localized / not localized messages that you noticed, but now have isolated them to the Jing library. In particular, I'm seeing messages from `com.thaiopensource.relaxing.pattern.resources`. This library appears to be using the default Locale, and knows nothing about the locale overrides. I can't think of a nice way around this issue and am hoping you can shed some light on it.

My biggest roadblock is time. When I submitted this PR several years ago, I frankly had plenty of time for side efforts. For the last year I've been the sole engineer at a startup and spare time is at a premium.

tofi86 commented:
> As for Jing's localization, I know we added the properties files in 8659d75, and it seems the localization logic is mostly done in the ErrorHandlerImpl. But not exactly sure where to go from there (I didn't dig deep).

They are basically only present in the repository so that Transifex sync can access the english base files. With the following commit e2796b240b4cd91ce0256a723c9ebe78cbdd5ace I remove them during the build process, so EpubCheck using the original Jing lib with english messages + additional non-english message files from our resources.
I don't think that they cause problems here.

As far as I understood, the main problem is, that there is no API to pass a Locale to Jing during validation.
So I think we should be following your proposal, Romain
> leave it as is and document the limitations

I'm fine with refactoring and/or joining the message classes at a later time.
At least we have perfect localization (options) integrated with our own message files and classes. Maybe we'll find another way to add this for Jing anytime in the future...

So @matthew-macgregor, as far as I understood you didn't push your latest changes yet as you stumbled upon the Jing class issues. Would you be okay with pushing your latest changes to this branch and rebasing one last time on top of `master` so that we can integrate the current version in 4.1?
Do you think you could do this within the next couple of days?

> My biggest roadblock is time. When I submitted this PR several years ago, I frankly had plenty of time for side efforts. For the last year I've been the sole engineer at a startup and spare time is at a premium.

I perfectly understand your situation and I'm deeply sorry that your contribution had to wait so long. But we greatly appreciate your contribution and really would like to see it in 4.1

rdeltour commented:
I updated and completed this PR, and pushed the changes to #911:

- rebased against the latest `master` branch
- store the current locale in a static thread-local variable (in [`LocaleHolder`](https://github.com/w3c/epubcheck/blob/24574a3178132a1a409a93acc9dc359d94293fca/src/main/java/com/adobe/epubcheck/messages/LocaleHolder.java)
- [monkey-patch Jing's `Localizer` class](https://github.com/w3c/epubcheck/blob/24574a3178132a1a409a93acc9dc359d94293fca/src/main/java/com/thaiopensource/util/Localizer.java) to use this statically-held locale to load the resource bundles.
- refactored the [`Messages`](https://github.com/w3c/epubcheck/blob/24574a3178132a1a409a93acc9dc359d94293fca/src/main/java/com/adobe/epubcheck/util/Messages.java) class to add a static `getInstance` method taking a class name in addition of a locale, to be able to load resource bundles from a different package. The underlying messages are now cached in a table keyed by both locale code and bundle name.
- refactor the CSS parser package to make it locale-aware
- add empty `_en` message properties files
- add a couple more localization tests.

Thanks again @matthew-macgregor for the contribution! 👍

rdeltour commented:
and also:

- introduced a `LocalizedReport` interface, extending `Report`, to avoid breaking the `Report` API. The `MasterReport` implementation uses this new `LocalizedReport` interface.

tofi86 commented:
Wow, you must have had a fun evening, @rdeltour 🤣 

Thanks a lot for finishing this PR 👍 💯 and thanks to @matthew-macgregor for the initial contribution 🥇!

tofi86 commented:
> ```
> ERROR (CSS-008) at "waldenfotos.epub/OEBPS/Walden_c1.xhtml" (line 10, col 3):
>    An error occurred while parsing the CSS: Die Zeichenkette ".quote" ist an dieser Stelle nicht erlaubt. Erwartet wird eine CSS-Eigenschaft.
> ```
> 
> This message appeared in my EPUB-Checker App after switching the language from `de_DE` to `en_US`.
> 
> The first part of the message is in english and is issued by the [EpubCheck message libs](https://github.com/matthew-macgregor/epubcheck/blob/localization-dev/src/main/resources/com/adobe/epubcheck/messages/MessageBundle.properties#L44) while the second part, which comes from the [Jing validation](https://github.com/matthew-macgregor/epubcheck/blob/localization-dev/src/main/resources/org/idpf/epubcheck/util/css/messages_de.properties#L10), is still in german.

Works perfect now! 👏
